src/spawn_stubs.c

```diff
@@ -488,6 +502,188 @@ CAMLprim value spawn_unix(value v_env,
+      break;
+    case FD:
+      e_error = posix_spawn_file_actions_addfchdir_np(&actions, info.cwd.fd);
+      if (e_error) {
+        e_function = "fchdir";
+        e_arg = Field(v_cwd, 0);
+        goto cleanup;
+      }
+      break;
+  }
+
+  for (int fd = 0; fd < 3; fd++) {
+    int tmp_fd = tmp_fds[fd] = safe_dup(info.std_fds[fd]);
+    if (tmp_fd == -1) {
+      e_error = errno;
+      e_function = "dup";
+      goto cleanup;
+    }
+
+    e_error = posix_spawn_file_actions_adddup2(&actions, tmp_fd, fd);
+    if (e_error) {
+      e_function = "dup2";
+      goto cleanup;
+    }
+
+    e_error = posix_spawn_file_actions_addclose(&actions, tmp_fd);
+    if (e_error) {
+      e_function = "close";
+      goto cleanup;
+    }
+  }
+
+  e_error = posix_spawn(&pid, info.prog,
+                        &actions, &attr,
+                        info.argv, info.env);
+  if (e_error) {
+    e_function = "posix_spawn";
+    goto cleanup;
+  }
+
+ cleanup:
+  for (int fd = 0; fd < 3; fd++)
+    if (tmp_fds[fd] > 2)
+      close(tmp_fds[fd]);
+
+  free_spawn_info(&info);
+  posix_spawnattr_destroy(&attr);
+  posix_spawn_file_actions_destroy(&actions);
+
+  if (e_function) {
+    unix_error(e_error, e_function, e_arg);
+  }
+
+  CAMLreturn(Val_int(pid));
+}
+
+#else
+
+CAMLprim value spawn_unix(value v_env,
+                          value v_cwd,
+                          value v_prog,
+                          value v_argv,
+                          value v_stdin,
+                          value v_stdout,
+                          value v_stderr,
+                          value v_use_vfork,
+                          value v_setpgid,
+                          value v_sigprocmask)
+{
+  CAMLparam4(v_env, v_cwd, v_prog, v_argv);
+  pid_t ret;
+  struct spawn_info info;
+  int result_pipe[2];
+  int cancel_state;
+  sigset_t sigset;
+  sigset_t saved_procmask;
+  struct subprocess_failure failure;
+  int got_error = 0;
+  int errno_after_forking = 0;
+  int status;
+
+  init_spawn_info(&info,
+                  v_env, v_cwd, v_prog, v_argv,
+                  v_stdin, v_stdout, v_stderr, v_setpgid);
+
+  caml_enter_blocking_section();
+  enter_safe_pipe_section();
+
+  /* Pipe used by the child to send errors to the parent. */
+  if (safe_pipe(result_pipe) == -1) {
+    int error = errno;
+    leave_safe_pipe_section();
+    caml_leave_blocking_section();
+    free_spawn_info(&info);
+    unix_error(error, "pipe", Nothing);
+  }
+
+  /* Block signals and thread cancellation. When using vfork, the
+     child might share the signal handlers.
+
+     It's not clear that we need the call to [pthread_setcancelstate],
+     but implementations of posix_spawn based on vfork are doing this.
+
+     For instance:
+     http://git.musl-libc.org/cgit/musl/tree/src/process/posix_spawn.c
+  */
+  pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
+  sigfillset(&sigset);
+  init_spawn_info_sigmask(&info, v_sigprocmask, &sigset, &saved_procmask);
```

Comment:
This bug seems to exist in the base revision, but I think we're unsafely accessing OCaml heap (`v_sigprocmask`) here from the blocking section.

Reply:
See 2b6c57b0818c9ac72237fcf52db088d02b526ed5 for (I think) a fix.